Add EIP: Private key deactivation and reactivation (#9193)
ethereum/EIPs: EIPS/eip-7851.md

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

### Parameters

| Constant                          | Value                |
|-----------------------------------|----------------------|
| `PRECOMPILE_ADDRESS`              | `0xTBD`              |
| `PRECOMPILE_GAS_COST`             | `5000` (tentative)   |

### Delegated code encoding

The deactivation status is encoded by appending or removing the `0x00` byte at the end of the delegated code. The transitions between two states are as follows:

- Active state: `0xef0100 || address`, the private key is active and can sign transactions.
- Deactivated state: `0xef0100 || address || 0x00`, the private key is deactivated and cannot sign transactions.

### Precompiled contract

A new precompiled contract is introduced at address `PRECOMPILE_ADDRESS`. For each call, it consumes `PRECOMPILE_GAS_COST` gas, and the precompiled contract executes the following steps:

> - The precompiled contract checks that the caller is an EOA with delegated code (i.e., its account code begins with the prefix `0xef0100`, as defined in [EIP-7702](./eip-7702)). If the code does not conform to the required prefix, the contract MUST terminate without making any state changes.

Reviewer: "Terminate" here means just "stop" the call to this precompile without making any state changes? (One could also "revert", thus the item put on stack when using the `CALL*` opcodes to this precompile would put 0 on stack if these conditions are not met (so no state changes are done?))

Author: Thanks for the clarification, the error case of precompile is updated as: "Returns a precompile contract error and consumes all provided gas" (refer to [0f2aaba](https://github.com/ethereum/EIPs/pull/9193/commits/0f2aabaad752324ad8bab51a123213ca290a1c10)).

The handling logic is consistent with other precompiles mentioned in https://www.evm.codes/precompiled. Also, consuming all gas for invalid calls (reducing DoS attack vectors) looks reasonable.

For the error return, I searched client implementation in [geth](https://github.com/ethereum/go-ethereum/blob/v1.14.12/core/vm/instructions.go#L756-L762) and found it handled outside by `CALL`, `STATICCALL`, `CALLCODE`, `DELEGATECALL`. When [precompiles return an error](https://github.com/ethereum/go-ethereum/blob/v1.14.12/core/vm/evm.go#L210-L244), it will be returned to `CALL`, and `CALL` will set return value as `0`. But formal documents would help clarify this for sure.

- The precompile determines the current state of the delegated code based on its byte length:
  - If the delegated code is 24 bytes (`0xef0100 || address || 0x00`), it removes the last byte (`0x00`), transitioning to the active state (`0xef0100 || address`).
  - If the delegated code is 23 bytes (`0xef0100 || address`), it appends `0x00`, transitioning to the deactivated state (`0xef0100 || address || 0x00`).

> - The updated delegated code is saved as the new account code for the EOA.

Reviewer: I think a rule should be added here. In case that we are updating the delegated code, but we are running in "static" mode (so a call chain after using `STATICCALL`) this should abort (revert, no state-changes).

Author: yes. you're right! I added the `STATICCALL` error case in this commit: [0f2aaba](https://github.com/ethereum/EIPs/pull/9193/commits/0f2aabaad752324ad8bab51a123213ca290a1c10)

---
eip: 7851
title: EOA private key deactivation/reactivation
description: Introduce a new precompiled contract for EOAs with delegated code to deactivate or reactivate private keys.
author: Liyi Guo (@colinlyguo)
discussions-to: https://ethereum-magicians.org/t/eip-7851-eoa-private-key-deactivation-reactivation/22344
status: Draft
type: Standards Track
category: Core
created: 2024-12-27

> requires: 20, 2612, 7701, 7702

Reviewer: ```suggestion
requires: 7702
```

I don't see how ERC-20, EIP 2612, and 7701 are required here. If you would require 7701 this means this EIP can only be active once EOF is also active.

Author: oh I see. thanks for the clarification here.

### Delegated code encoding

The deactivation status is encoded by appending or removing the `0x00` byte at the end of the delegated code. The transitions between two states are as follows:

- Active state: `0xef0100 || address`, the private key is active and can sign transactions.
- Deactivated state: `0xef0100 || address || 0x00`, the private key is deactivated and cannot sign transactions.

### Precompiled contract

A new precompiled contract is introduced at address `PRECOMPILE_ADDRESS`. It costs `POINT_EVALUATION_PRECOMPILE_GAS` and executes the following logic:

- Returns a precompile contract error, consumes all gas, and no state changes are made if:
  - Gas is insufficient.
  - Called via `STATICCALL` (i.e. in a read-only context).
  - Caller is not an EOA with delegated code (prefix `0xef0100` as per [EIP-7702](./eip-7702)).
- Updates caller's delegated code based on length:
  - 23 bytes (`0xef0100 || address`): Appends `0x00` to deactivate private key authorization.
  - 24 bytes (`0xef0100 || address || 0x00`): Removes last byte `0x00` to activate private key authorization.
- Saves updated code as caller's new account code.

> ### Transaction validation

Author: reported.

If the account is verified as an EOA with a delegated code (begins with the prefix `0xef0100`), the following validations MUST be performed:

> - Transactions signed by the private key MUST be rejected if the delegated code is in the deactivated state (i.e., `24` bytes long).

Reviewer: I think extra clarification here is necessary. What does "rejected" mean? I am assuming that "rejected" here means that the transactions which do not meet this rule can be included in the block, however any state changes like EVM execution, publishing blobs, or authorizing (EIP-7702) are not done and instead the transaction immediately exits. However, this does mean that these state changes are still done:

- Nonce update
- Paying for gas (so balance update)

Also, for this rejection: is the entire gas limit consumed or only the intrinsic gas (which includes for instance the calldata fee)

Author: added clarifications in this commit: [4ff1e76](https://github.com/ethereum/EIPs/pull/9193/commits/4ff1e76241ae136f4f2c79ed6c0a03d1e28bee46).

also removing the basic 21000 gas discussion, since it seems to be unrelated now.

- Any [EIP-7702](./eip-7702) authorization from an authority with deactivated delegated code MUST be considered invalid and skipped.

### Gas Cost

> No changes to the base transaction gas cost (`21000`) are required, as the additional valid check for the deactivation status is minimal. Thus it is reasonable to consider the overhead covered by the base gas cost.

Reviewer: The check if the current account is deactivated actually is more expensive than a base transaction due to the nature of the state trie. To read an account from the trie, one gets these items:

- Nonce
- Balance
- `codeHash`
- `storageHash`

In order to read the actual code (to figure out if the account is delegated and thus to also perform the private key deactivation check) we have to actual read from the database again (to find what code `codeHash` points to). Since this is a disk read this is expensive and since this is now necessary for all transactions this might be a reason to increase the base transaction cost.

Author: it makes sense. I missed this (`code` read cost).

In the first draft before opening this pr, the proposal plans to add a new bool (1 byte) field in the account state, so that it can be read together with `nonce`, `balance`, `codeHash`, and `storageHash`.

this may add some complexities in backward compatibility discussions while keeping the basic cost of transactions cheaper. do you think it makes sense to change the implementation of the draft to this method?

Reviewer: @colinlyguo This extra field in state will make this rather complex, because you are adding an extra field to the account state. This also means there will be logic introduced when enabling this EIP: each time if an account is hit, there will be logic to see if this account is "upgraded" to have this boolean in state. If this is not the case, then it could either add this flag to the state, or it could also do it only when necessary (so, if a transaction is sent, there will first be a check if the private key is deactivated. To check, first see if there is a boolean field at all: if this is not the case, then we can know for sure that this is an "old" account state and the private key is thus active. We can then add this field to the state. But this mechanism of checking and adding these fields will make the test vectors for this very big and it will also add a lot of somewhat boilerplate "upgrade" code to ensure the old accounts will update to the new ones (with this private key enabled flag). I would not recommend it.

I do recall in the past there have been discussions of using the `nonce` field. One could, for instance, use the highest bit of the nonce for this flag. One could use this flag to see if the accounts private key is disabled. What is handy here is that the nonce data is already in the account, so no extra disk lookups are necessary (when compared to putting it in the code).

However some nonce-limiting EIPs are around: https://eips.ethereum.org/EIPS/eip-3338, https://eips.ethereum.org/EIPS/eip-4803, putting a flag in nonce would likely throw away the benefits/rationales of these EIPs (4803 ensures that it can be decoded as int).

Author: agree. introduced a new field in the account state will add a lot of unnecessary "upgrade" code, which should be avoided.

adding a flag in `nonce` looks like a good idea.

for [EIP-3338](https://eips.ethereum.org/EIPS/eip-3338) which limits the nonce <= 2^52, using the highest bit of 2^52 seems not to make a large difference: (i) the number is still large, 2^51 = 2.25 * 10^15. (ii) compared with 2^64-1, it's also quite small as 2^52.

To limit the range of future nonce, can move the "deactivated bit" based on the length of the nonce after an upgrade.

while this will also introduce a problem (maybe?) that historical nonce in some reverted transactions (those included but reverted since the EOA is deactivated) will exceed future limited value ("languages that only have floating point number support" may not be happy to see that).

For this, I have come up with 2 solutions:
(i) reject transactions in the "signature validation phase" if the highest bit of its nonce is 1.
(ii) encode the "deactivated bit" in the lowest bit of the "nonce" field.

Reviewer 2: Ok

Author: for complementary, I found that nodes already load code during transaction validation due to EIP-3607 and EIP-7702, so adding a length check is negligible.

code loading in geth implementation: https://github.com/ethereum/go-ethereum/blob/0e1a19da76d1761fdb3db78112a9d0e3571d43a4/core/state_transition.go#L323

and transaction validation in `txpool` still needs getting `code`, which is comparable to `nonce` and `balance` checks.

---
eip: 7851

> title: EOA private key deactivation/reactivation

Reviewer 3: ```suggestion
title: EOA delegation deactivation/reactivation
```

> title: EOA private key deactivation/reactivation

Reviewer 3: ```suggestion
title: Delegated EOA's private key deactivation/reactivation
```
may be doesn't fit the title length constraint...

Author: yes. While the added `delegated` seems necessary, what about `deactivate/reactivate of delegated EOA's key`?

Reviewer 3: yes sounds good

> description: Introduce a new precompiled contract for EOAs with delegated code to deactivate or reactivate private keys.

Reviewer 3: ```suggestion
description: Introduce a new precompiled contract for EOAs with delegated code to deactivate or reactivate delegation.
```

author: Liyi Guo (@colinlyguo)
discussions-to: https://ethereum-magicians.org/t/eip-7851-eoa-private-key-deactivation-reactivation/22344
status: Draft
type: Standards Track
category: Core
created: 2024-12-27
requires: 7702
---

## Abstract

> This EIP introduces a precompiled contract that enables Externally Owned Accounts (EOAs) with delegated control to smart contracts via [EIP-7702](./eip-7702) to deactivate or reactivate their private keys. This design does not require additional storage fields or account state changes. By leveraging delegated code, reactivation can be performed securely through mechanisms such as social recovery.

Reviewer 3: ```suggestion
This EIP introduces a precompiled contract that enables Externally Owned Accounts (EOAs) with delegated control to smart contracts via [EIP-7702](./eip-7702) to deactivate or reactivate their delegation. This design does not require additional storage fields or account state changes. By leveraging delegated code, reactivation can be performed securely through mechanisms such as social recovery.
```

## Rationale

### Using a precompiled contract

Alternative methods for implementing this feature include:

- Adding a new transaction type: Introducing a new transaction type could provide a mechanism to deactivate and reactivate EOA private keys. However, reactivating the private key would rely on a delegated contract as the authorizer, which complicates defining the rules for the new transaction type.
- Deploying a regular smart contract: A regular deployed contract could track the deactivated status of each address. However, invoking this contract, which executes bytecode and accesses storage, during deactivation and reactivation, would be more expensive than using a precompiled contract. Additionally, this approach "leaks" the design of a (widely used) programming language into the Ethereum core protocol. While it poses no obvious security risks, it is not ideal from a design perspective.

### In-protocol reactivation

> This approach ensures maximum compatibility with future migrations. EOAs can reactivate their private keys, delegate their accounts to an [EIP-7701](./eip-7701) contract, and then deactivate their private keys again. This avoids the limitations of contract upgrades. e.g. to remove legacy proxy contracts (reducing gas overhead) when upgrading to EOF contracts, one can reactivate the EOA and redelegate it to an EOF proxy contract.

Reviewer 4: I see no better alternative, but this definitely sounds risky. Once a privkey has been deactivated most users would not take care to keep it secure. If it's leaked an attacker could use it after reactivation.

A powerful solution would be to enable delegation through a precompile. Then the account code could implement the required checks before "upgrading" its own code.

> This approach ensures maximum compatibility with future migrations. EOAs can reactivate their private keys, delegate their accounts to an [EIP-7701](./eip-7701) contract, and then deactivate their private keys again. This avoids the limitations of contract upgrades. e.g. to remove legacy proxy contracts (reducing gas overhead) when upgrading to EOF contracts, one can reactivate the EOA and redelegate it to an EOF proxy contract.

Reviewer 4: > EOAs can reactivate their private keys

The code cannot know if its privkey is currently deactivated or not, right? (Since `extcodehash` only returns the delegation prefix.) Would this be an issue?

Author: I think it's OK not to know the specific status, each time the contract invokes the precompile, it just toggles the status, i.e. from `deactivated` to `activated`, or from `activated` to `deactivated`.

        # Check static call
        if static:
            return b"cannot call in static context", 0

        # Get and validate caller's code
        code = state_db.get_code(caller)
        if not code.startswith(self.DELEGATED_CODE_PREFIX):
            return b"invalid delegated code prefix", 0

        # Update code based on length
        if len(code) == 23:  # Active state
            state_db.set_code(caller, code + b"\x00")  # Deactivate
        elif len(code) == 24:  # Deactivated state
            state_db.set_code(caller, code[:-1])  # Activate
        else: # Although this is not possible, it's added for completeness
            return b"invalid code length", 0

        return b"", gas - self.GAS_COST

class StateDB:
    """Simplified state database, omitting other account fields"""
    def __init__(self):
        self.accounts = {}

    def get_code(self, addr):
        return self.accounts.get(addr, {}).get("code", b"")

    def set_code(self, addr, value):
        if addr not in self.accounts:
            self.accounts[addr] = {}
        self.accounts[addr]["code"] = value
```

## Security Considerations

### Additional checks during transaction validation

> The deactivation status is determined by checking the length of the delegated code. This check introduces an additional storage read and check comparable to account state read and checks, such as nonce validation. It is integrated into the transaction validity check of the transaction pool and consensus rules, ensuring that invalid transactions are filtered out before propagation and execution.

Reviewer 4: > This check introduces an additional storage read

Since we read the account anyway for the nonce, I don't think this requires an additional read, so there's no additional workload for the txpool.

Reviewer: There is an extra storage read, because the code itself is not directly stored in the account: only the code hash is stored in the account. So an extra storage read is necessary to get the actual code of the account (and not the hash).
